### PR TITLE
feat(freespace_planning_algorithms): implement support for multiple goal candidates in A star planner

### PR DESCRIPTION
**Pull request by mkquda**
## Description
Free space planner is sometimes used by goal planner to find a path for pull over. In such cases, goal planner will generate multiple goal candidates and execute freespace planner for each candidate until a path is found ([see here](https://autowarefoundation.github.io/autoware.universe/main/planning/behavior_path_planner/autoware_behavior_path_goal_planner_module/#goal-search)).

In some cases it can take a long time to generate the path (when seach fails for first candidate(s)), as the planner repeats the search process for each goal.

A more efficient approach is to provide multiple goal candidates to the A star planner in one go, and try to find a path to any of the goals in one search process.

## Changes
Includes changes from preceding [PR](https://github.com/autowarefoundation/autoware.universe/pull/8091)

Extend A star implementation to accept a vector of goal poses
The goal poses will be filtered, and only feasible goals are kept
The first feasible goal will be considered as the primary goal

In case of forward search configuration:
- primary goal will be set as the goal pose
- heuristic cost estimate will be computed w.r.t to primary goal
- search process will terminate once any of the goal candidates is reached

In case of backward search configuration:
- all goal candidates will be set as search start pose(s)
- start pose will be set as the search goal
- open list will be initialized with all goal candidates
- to account for candidates priority, an initial cost offset will be assigned to each start node
- search process will terminate once the search goal (start pose) is reached

## Related links

**Preceding PR's:**

- https://github.com/autowarefoundation/autoware.universe/pull/8068
- https://github.com/autowarefoundation/autoware.universe/pull/8089
- https://github.com/autowarefoundation/autoware.universe/pull/8091

## How was this PR tested?
tested on psim.

## Notes for reviewers

This PR includes changes from a preceding [PR](https://github.com/autowarefoundation/autoware.universe/pull/8091). so please review the PR's in order.

## Interface changes

None.

## Effects on system behavior

A start planner accepts multiple goal candidates

### CODE REVIEW

**Review by armaganarsln — CHANGES_REQUESTED** [review]
Can you please test with this issue: 
- https://github.com/autowarefoundation/autoware.universe/issues/8342

**Review by armaganarsln — APPROVED** [review]

**Comment by xmfcx** [comment]
Could you rebase to the latest main?

File: planning/autoware_freespace_planning_algorithms/src/astar_search.cpp [hunk 1]
```diff
@@ -174,25 +237,21 @@ void AstarSearch::setCollisionFreeDistanceMap()
...
+void AstarSearch::setStartNode(const double cost_offset)
 {
   const auto index = pose2index(costmap_, start_pose_, planner_common_param_.theta_size);
-
-  if (detectCollision(index)) return false;
-
   // Set start node
   AstarNode * start_node = &graph_[getKey(index)];
-  start_node->set(start_pose_, 0.0, estimateCost(start_pose_, index), 0, false);
+  double initial_cost = estimateCost(start_pose_, index) + cost_offset;
```
Comment (kosuke55) [inline planning/autoware_freespace_planning_algorithms/src/astar_search.cpp:245]:
```suggestion
  const double initial_cost = estimateCost(start_pose_, index) + cost_offset;
```

File: planning/autoware_freespace_planning_algorithms/src/astar_search.cpp [hunk 2]
```diff
@@ -242,7 +301,9 @@ bool AstarSearch::search()
 void AstarSearch::expandNodes(AstarNode & current_node, const bool is_back)
 {
   const auto current_pose = node2pose(current_node);
-  double distance = getExpansionDistance(current_node) * (is_back ? -1.0 : 1.0);
+  double direction = is_back ? -1.0 : 1.0;
```
Comment (kosuke55) [inline planning/autoware_freespace_planning_algorithms/src/astar_search.cpp:304]:
```suggestion
  const double direction = is_back ? -1.0 : 1.0;
```

File: planning/autoware_freespace_planning_algorithms/src/astar_search.cpp [hunk 3]
```diff
@@ -242,7 +301,9 @@ bool AstarSearch::search()
...
+  if (is_backward_search_) direction *= -1.0;
+  double distance = getExpansionDistance(current_node) * direction;
```
Comment (kosuke55) [inline planning/autoware_freespace_planning_algorithms/src/astar_search.cpp:306]:
```suggestion
  const double distance = getExpansionDistance(current_node) * direction;
```

File: planning/autoware_freespace_planning_algorithms/src/astar_search.cpp [hunk 4]
```diff
@@ -333,46 +411,65 @@ void AstarSearch::setPath(const AstarNode & goal_node)
...
+  }
+
+  const auto interpolate = [this, &waypoints, &pose](const AstarNode & node) {
     if (node.parent == nullptr || !astar_param_.adapt_expansion_distance) return;
     const auto parent_pose = node2pose(*node.parent);
     const double distance_2d = calcDistance2d(node2pose(node), parent_pose);
     const int n = static_cast<int>(distance_2d / min_expansion_dist_);
     for (int i = 1; i < n; ++i) {
-      const double dist = ((distance_2d * i) / n) * (node.is_back ? -1.0 : 1.0);
+      double dist = ((distance_2d * i) / n) * (node.is_back ? -1.0 : 1.0);
```
Comment (kosuke55) [inline planning/autoware_freespace_planning_algorithms/src/astar_search.cpp:433]:
```suggestion
      const double dist = ((distance_2d * i) / n) * (node.is_back ? -1.0 : 1.0);
```

File: planning/autoware_freespace_planning_algorithms/src/astar_search.cpp [hunk 5]
```diff
@@ -382,26 +479,65 @@ bool AstarSearch::isGoal(const AstarNode & node) const
...
+      return false;
+    }
+
+    const auto angle_diff =
+      autoware::universe_utils::normalizeRadian(tf2::getYaw(relative_pose.orientation));
+    if (std::abs(angle_diff) > goal_angle) {
+      return false;
+    }
+
+    bool is_set_shifted_goal_pose =
```
Comment (kosuke55) [inline planning/autoware_freespace_planning_algorithms/src/astar_search.cpp:511]:
```suggestion
    const bool is_set_shifted_goal_pose =
```

**Review by kosuke55 — APPROVED** [review]
thanks a lot!

**Comment by NilaySener** [comment]
Hi @mkquda ,
As we discussed in the previous Software and Planning-Control working group, we planned to test your changes on our autonomous test vehicle within the campus. Here you can find more information about our testing environment:
- https://github.com/autowarefoundation/autoware.universe/issues/8114
- PCD Map File: Download [here.](https://drive.google.com/file/d/1Hqdh2hHIyNYeRRvY23-tN_TBcA_rq_zS/view?usp=sharing)

To begin, I initially tested your PR locally using the planning simulator. However, I encountered issues where the vehicle failed to complete certain parking scenarios during testing.

### **1. L Parking**
- Lanelet2 Map: Download [here](https://drive.google.com/file/d/11X2qR6Y53cbz2fjRjUNekJgqcb5NDx0l/view?usp=sharing).
- Initial Pose: 
```
ros2 topic pub /planning/mission_planning/goal geometry_msgs/msg/PoseStamped "{header: {stamp: {sec: 0, nanosec: 0}, frame_id: 'map'}, pose: {position: {x: 58890.332031 , y: 42795.171875, z: 73.713248}, orientation: {x: -0.021747, y: -0.000537, z: -0.999459, w: 0.024657}}}"
```
- Goal Pose:
 ```
ros2 topic pub /planning/mission_planning/goal geometry_msgs/msg/PoseStamped "{header: {stamp: {sec: 0, nanosec: 0}, frame_id: 'map'}, pose: {position: {x: 58863.644531, y: 42799.714844, z: 0.0}, orientation: {x: 0.0, y: 0.0, z: 0.696914, w: 0.717155}}}"
```

- Output error:
```
[component_container_mt-34] [INFO 1724317099.616951061] [planning.scenario_planning.parking.freespace_planner]: Can't find goal: Invalid start or goal pose (planTrajectory() at /home/nilay/workspace/ups/autoware/src/universe/autoware.universe/planning/autoware_freespace_planner/src/autoware_freespace_planner/freespace_planner_node.cpp:576)
[component_container-31] [ERROR 1724317099.648966319] [planning.scenario_planning.velocity_smoother]: input trajectory size must > 1. Skip computation. (checkData() at /home/nilay/workspace/ups/autoware/src/universe/autoware.universe/planning/autoware_velocity_smoother/src/node.cpp:417)

```
https://github.com/user-attachments/assets/4d04df4f-25fe-4d0e-9975-d5aff900e671


### **2. L Parking: (modified lanelet2 file)**
In this scenario, the parking lot (free space area) has been increased to solve the trajectory problem that the vehicle will produce for parking (Moving away from real life measurements). Here, I would like to point out that the vehicle **does not use** the space I increased when generating the parking tarjectory.


- Lanelet2 Map: Download [here.](https://drive.google.com/file/d/1rzrYJ375Kl8EZwae4ca_BRAaMUcCQ8hv/view?usp=sharing)
- Initial Pose:
```
ros2 topic pub /initialpose geometry_msgs/msg/PoseWithCovarianceStamped "{header: {stamp: {sec: 0, nanosec: 0}, frame_id: 'map'}, pose: {pose: {position: {x: 58890.117188, y: 42795.402344, z: 72.721600}, orientation: {x: 0.021813, y: -0.000139, z: 0.999742, w: 0.006372}}, covariance: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]}}"
``` 

- Goal Pose: 
```
ros2 topic pub /planning/mission_planning/goal geometry_msgs/msg/PoseStamped "{header: {stamp: {sec: 0, nanosec: 0}, frame_id: 'map'}, pose: {position: {x: 58863.953125, y: 42799.664062, z: 0.0}, orientation: {x: 0.0, y: 0.0, z: 0.686601, w: 0.727034}}}"
```

https://github.com/user-attachments/assets/6730f3a2-5be7-44f8-8625-67e0df6790ce

### **3. L Parking Next to Parked Vehicle: (modified lanelet2 file)**

- Lanelet2 Map: Same file as **2. L Parking**
- Initial Pose: 
```
ros2 topic pub /initialpose geometry_msgs/msg/PoseWithCovarianceStamped "{header: {stamp: {sec: 0, nanosec: 0}, frame_id: 'map'}, pose: {pose: {position: {x: 58889.300781, y: 42795.335938, z: 72.672600}, orientation: {x: 0.021806, y: -0.000481, z: 0.999519, w: 0.022029}}, covariance: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]}}"
```

- Goal Pose:
 ```
ros2 topic pub /planning/mission_planning/goal geometry_msgs/msg/PoseStamped "{header: {stamp: {sec: 0, nanosec: 0}, frame_id: 'map'}, pose: {position: {x: 58863.863281, y: 42799.839844, z: 0.0}, orientation: {x: 0.0, y: 0.0, z: 0.700313, w: 0.713835}}}"
```


https://github.com/user-attachments/assets/d0e18499-67dc-4499-9441-6e33cf518e44

---

### **Test Summary**

Several scenarios involving "L Parking" were evaluated within a simulated environment are given above. As can be seen from the videos,  the vehicle encountered challenges in completing the parking maneuvers. However, after the changes made to the free space area in the lanelet2 map file, the vehicle was able to produce a parking trajectory. However, since we produce the map according to real life dimensions, this is not a safe method. Despite these changes, the vehicle is still struggling when parking next to a parked vehicle.
These findings highlight the need for further adjustments to both the parking logic and map configuration to ensure reliable operation in real-world scenarios.

**Comment by mkquda** [comment]
@NilaySener 

Thank you for your test and report, much appreciated!

I understand that even with the recent improvements the parking feature is still lacking, and there is still room for improvement.
Let me highlight that the recent changes were made to improve the core free space planning algorithm. Improvements to parking feature it self as a whole is still needed, and should be worked on in the future.

With regards to the specific samples you shared:

**1. L Parking**

In this case the Ego footprint (with margin) at the goal pose is very close to the edge of the free space area, so there is a high chance it overlaps with blocked occupancy grid cells and the goal pose will be considered invalid on the freespace planning algorithm side. it is necessary to have some buffer between the edge of the free space area and the edge of the parking lots where the ego is expected to park.
(for your reference, the image below shows how close the ego would be to edge, at the provided goal pose)
Screenshot from 2024-08-23 08-27-30

**3. L Parking Next to Parked Vehicle: (modified lanelet2 file)**

I understand that the parking spaces are drawn to real life dimensions, but with the current safety buffers we have it would be difficult to park next to another vehicle without any extra space between the parking slots.
(for your reference, image below shows the occupancy grid configuration in this case)
Screenshot from 2024-08-23 08-29-04


I removed the safety margins and did the following tests using the map you provided.

**L Parking Next to Vehicle**

https://github.com/user-attachments/assets/3da3bcec-499d-424b-af33-c1d42cfef8e7

**L Parking Between 2 Vehicles**

https://github.com/user-attachments/assets/a4a419c6-716f-418f-9482-b3042e15818c

**Pull request merged**